InterruptClient.rb

```diff
@@ -0,0 +1,230 @@
+      to_read = IO.select([@client, STDIN]) # check what's ready to read: socket, or terminal input
+      to_read[0].each { |ios|
+        if ios == @client
+          msg, sender = ios.recvfrom(MAX_MSG_LENGTH)
+          handle_msg(msg, sender)
+        elsif ios.tty? # ios comes from terminal
+          input = STDIN.getc
+          handle_key(input)
+        end
+      }
+    end
+  end
+
+  def handle_msg(msg, sender)
+    msg = parse_msg(msg, sender)
+    return if msg.nil?
+
+    case msg['type']
+    when 'chat'
+      time = msg['time']
+      return if time < @latest_chat
+
+      @latest_chat = time
+      chat_array = msg['body']
+      names_array = msg['names']
+
+      names_string, names_length = names_data(names_array)
+
+      string = Console.left(chat_array.length) + Console.up(10) +
+          names_string +
+          Console.left(names_length) + Console.down(10) +
+          chat_string(chat_array)
+
+      print string
+    end
+  end
+
+  def chat_string(chat_array)
+    chat_array.inject('') { |string, item|
+        text, color_code = item
+        color = color_code.nil? ? nil : COLORS[color_code]
+
+        string + Console.color(color, text)
+      }
+  end
+
+  def names_data(names_array)
+    names_string = ''
+    names_length = 0 # names_string may have color encoding chars, so we can't just check its length
+
+    names_array.each {|item|
+      nickname, color_code, emph = item
+      names_length += nickname.length + 2
+
+      nickname = color_code.nil? ? nickname : Console.color(COLORS[color_code], nickname)
+      nickname = emph ? Console.emph(nickname) : nickname
+      names_string += " #{nickname} "
+      }
+
+    return [names_string, names_length]
+  end
+
+  # disregard eg return, delete, backspace keys presses
+  # only pass on to server word, punctuation, or space characters
+  # TODO: strip out two chars following '\e'
+  def handle_key(input)
+    if input == CMD_QUIT
+      bye
+    elsif not /^[[[:word:]][[:punct:]] ]$/.match(input).nil?
+      send_msg(msg_chat(input))
+    end
+  end
+
+  # return parsed msg if correct format and comes from Interrupt server
+  # return nil otherwise
+  def parse_msg(data, sender)
+    return nil if not is_server?(sender)
+
+    begin
+      msg = JSON.parse(data)
+      msg = check_fields(msg)
+    rescue
+      nil
+    end
+  end
+
+  def is_server?(sender)
+    port = sender[1]
+    host = sender[2]
+
+    (port == @server_port) && (host == @server_host)
+  end
+
+  # return nil if fields wrong, otherwise return msg
+  def check_fields(msg)
+    basic_fields = (msg.has_key?('type') && msg.has_key?('time'))
+    return nil if not basic_fields
+
+    case msg['type']
+    when 'chat'
+      msg if (msg.has_key?('body') && msg.has_key?('names'))
+    when 'ack'
+      msg
+    else
+      nil
+    end
+  end
+
+
+  def send_msg(msg)
+    msg['time'] = Time.now.to_f.to_s
+    json = msg.to_json
+    @client.send(json, 0)
+  end
+
+  def msg_chat(body)
```

Comment:
whenever I find myself writing multiple methods that start with the same thing, e.g. `msg_` it smells like it's supposed to be another object. I recommend having `msg_connect`, `msg_chat`, `msg_quit` in a new object.